Add a kotlin panache guide (#9611)
quarkusio/quarkus: docs/src/main/asciidoc/hibernate-orm-panache-kotlin.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Quarkus - Simplified Hibernate ORM with Panache and Kotlin

include::./attributes.adoc[]
:config-file: application.properties

Hibernate ORM is the de facto standard JPA implementation and is well-known in the Java ecosystem.  Panache offers a
new layer atop this familiar framework.  This guide will not dive in to the specifics of either as those are already
covered in the link:hibernate-orm-panache.adoc[Panache guide].  In this guide, we'll cover the Kotlin specific changes
needed to use Panache in your Kotlin-based Quarkus applications.

== First: an example

> As we saw in the Panache guide, Panache allows us to extend the functionality in our entities and DAOs with some automatically

Reviewer: ```suggestion
As we saw in the Panache guide, Panache allows us to extend the functionality in our entities and repositories (also known as DAOs) with some automatically
```

provided functionality.  When using Kotlin, the approach is very similar to what we see in the Java version with a slight
change or two.  To Panache-enable your entity, you would define it something like:

[source,kotlin]
----
@Entity
class Person: PanacheEntity {
    lateinit var name: String
    lateinit var birth: LocalDate
    lateinit var status: Status
}
----

> As you can see our entities remain simple there is, however, a slight difference from the Java version.  The Kotlin

Reviewer 2: ```suggestion
As you can see our entities remain simple, there is however, a slight difference from the Java version.  The Kotlin
```

> language doesn't support the notion of static methods in quite the same way as Java does.  Instead, we use the

Reviewer: ```suggestion
language doesn't support the notion of static methods in quite the same way as Java does.  Instead, we must use a
```

> `companion object`:

Reviewer 2: ```suggestion
[companion object](https://kotlinlang.org/docs/tutorials/kotlin-for-py/objects-and-companion-objects.html#companion-objects):
```

[source,kotlin]
----
@Entity
class Person : PanacheEntity {
    companion object: PanacheCompanion<Person> {  // <1>
        fun findByName(name: String) = find("name", name).firstResult()
        fun findAlive() = list("status", Status.Alive)
        fun deleteStefs() = delete("name", "Stef")
    }

    lateinit var name: String  // <2>
    lateinit var birth: LocalDate
    lateinit var status: Status
}
----
<1> The companion object holds all the methods not related to a specific instance allowing for general management and
querying bound to a specific type.
<2> Here there are options, but we've chosen the `lateinit` approach.  This allows us to declare these fields as non-null
knowing they will be properly assigned either by the constructor (not shown) or by hibernate loading data from the
database.

> NOTE: Note that these types differ from the Java types mentioned in those tutorials.  For kotlin support, all the Panache

Reviewer: ```suggestion
NOTE: These types differ from the Java types mentioned in those tutorials.  For kotlin support, all the Panache
```

types will be found in the `io.quarkus.hibernate.orm.panache.kotlin` package.  This subpackage allows for the distinction
between the Java and Kotlin variants and allows for both to be used unambiguously in a single project.

> In the Kotlin version, we've simply moved the bulk of the link:https://www.martinfowler.com/eaaCatalog/activeRecord.html[`entity pattern`]

Reviewer: ```suggestion
In the Kotlin version, we've simply moved the bulk of the link:https://www.martinfowler.com/eaaCatalog/activeRecord.html[`active record pattern`]
```

For some reason, the Hibernate ORM with Panache guide seems to prefer this term, even though I don't agree with it.

functionality to the `companion object`.  Apart from this slight change, we can then work with our types in ways that map easily
from the Java side of world.

== Using the repository pattern

=== Defining your entity

When using the repository pattern, you can define your entities as regular JPA entities.
[source,kotlin]
----
@Entity
class Person {
    @Id
    @GeneratedValue
    var id: Long? = null;
    lateinit var name: String
    lateinit var birth: LocalDate
    lateinit var status: Status
}
----

=== Defining your repository

When using Repositories, you get the exact same convenient methods as with the active record pattern, injected in your Repository,

> by making them implements `PanacheRepository`:

Reviewer 2: ```suggestion
by making them implement `PanacheRepository`:
```

[source,kotlin]
----
class PersonRepository: PanacheRepository<Person> {
     fun findByName(name: String) = find("name", name).firstResult()
     fun findAlive() = list("status", Status.Alive)
     fun deleteStefs() = delete("name", "Stef")
}
----

All the operations that are defined on `PanacheEntityBase` are available on your repository, so using it
is exactly the same as using the active record pattern, except you need to inject it:

[source,kotlin]
----
@Inject
lateinit var personRepository: PersonRepository

@GET
fun count() = personRepository.count()
----

=== Most useful operations

Once you have written your repository, here are the most common operations you will be able to perform:

[source,kotlin]
----
// creating a person

> var person: Person? = Person()

Reviewer: Isn't it a bit weird to declare it as a nulltable type here? I guess we re-use the variable later, but still…

Author: Yeah, the only reason is the reuse below there.  But I'll tweak it.
